// spinal-env-viewer-context-geographic-service/build/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
const CONTEXT_TYPE = "geographicContext";
const SITE_TYPE = "geographicSite";
const BUILDING_TYPE = "geographicBuilding";
const FLOOR_TYPE = "geographicFloor";
const ZONE_TYPE = "geographicZone";
const ROOM_TYPE = "geographicRoom";
const EQUIPMENT_TYPE = "BIMObject";
const REFERENCE_TYPE = "geographicReference";

const SITE_RELATION = "hasGeographicSite";
const BUILDING_RELATION = "hasGeographicBuilding";
const FLOOR_RELATION = "hasGeographicFloor";
const ZONE_RELATION = "hasGeographicZone";
const ROOM_RELATION = "hasGeographicRoom";
const EQUIPMENT_RELATION = "hasBimObject";
const REFERENCE_RELATION = "hasReferenceObject";

const SITE_REFERENCE_CONTEXT = ".SiteContext";
const BUILDING_REFERENCE_CONTEXT = ".BuildingContext";
const FLOOR_REFERENCE_CONTEXT = ".FloorContext";
const ZONE_REFERENCE_CONTEXT = ".ZoneContext";
const ROOM_REFERENCE_CONTEXT = ".RoomContext";

const GEOGRAPHIC_TYPES = Object.freeze([SITE_TYPE, BUILDING_TYPE, FLOOR_TYPE, ZONE_TYPE, ROOM_TYPE]);

const GEOGRAPHIC_TYPES_ORDER = Object.freeze([CONTEXT_TYPE, SITE_TYPE, BUILDING_TYPE, FLOOR_TYPE, ZONE_TYPE, ROOM_TYPE, EQUIPMENT_TYPE]);

const GEOGRAPHIC_RELATIONS = Object.freeze([SITE_RELATION, BUILDING_RELATION, FLOOR_RELATION, ZONE_RELATION, ROOM_RELATION, EQUIPMENT_RELATION]);

const GEOGRAPHIC_RELATIONS_ORDER = Object.freeze([SITE_RELATION, BUILDING_RELATION, FLOOR_RELATION, ZONE_RELATION, ROOM_RELATION, EQUIPMENT_RELATION]);

const MAP_TYPE_RELATION = Object.freeze(new Map([[SITE_TYPE, SITE_RELATION], [BUILDING_TYPE, BUILDING_RELATION], [FLOOR_TYPE, FLOOR_RELATION], [ZONE_TYPE, ZONE_RELATION], [ROOM_TYPE, ROOM_RELATION], [EQUIPMENT_TYPE, EQUIPMENT_RELATION]]));

const MAP_RELATION_TYPE = Object.freeze(new Map([[SITE_RELATION, SITE_TYPE], [BUILDING_RELATION, BUILDING_TYPE], [FLOOR_RELATION, FLOOR_TYPE], [ZONE_RELATION, ZONE_TYPE], [ROOM_RELATION, ROOM_TYPE], [EQUIPMENT_RELATION, EQUIPMENT_TYPE]]));

exports.CONTEXT_TYPE = CONTEXT_TYPE;
exports.SITE_TYPE = SITE_TYPE;
exports.BUILDING_TYPE = BUILDING_TYPE;
exports.FLOOR_TYPE = FLOOR_TYPE;
exports.ZONE_TYPE = ZONE_TYPE;
exports.ROOM_TYPE = ROOM_TYPE;
exports.GEOGRAPHIC_TYPES = GEOGRAPHIC_TYPES;
exports.EQUIPMENT_TYPE = EQUIPMENT_TYPE;
exports.GEOGRAPHIC_TYPES_ORDER = GEOGRAPHIC_TYPES_ORDER;
exports.SITE_RELATION = SITE_RELATION;
exports.BUILDING_RELATION = BUILDING_RELATION;
exports.FLOOR_RELATION = FLOOR_RELATION;
exports.ZONE_RELATION = ZONE_RELATION;
exports.ROOM_RELATION = ROOM_RELATION;
exports.GEOGRAPHIC_RELATIONS = GEOGRAPHIC_RELATIONS;
exports.EQUIPMENT_RELATION = EQUIPMENT_RELATION;
exports.GEOGRAPHIC_RELATIONS_ORDER = GEOGRAPHIC_RELATIONS_ORDER;
exports.MAP_TYPE_RELATION = MAP_TYPE_RELATION;
exports.MAP_RELATION_TYPE = MAP_RELATION_TYPE;
exports.REFERENCE_TYPE = REFERENCE_TYPE;
exports.REFERENCE_RELATION = REFERENCE_RELATION;
exports.SITE_REFERENCE_CONTEXT = SITE_REFERENCE_CONTEXT;
exports.BUILDING_REFERENCE_CONTEXT = BUILDING_REFERENCE_CONTEXT;
exports.FLOOR_REFERENCE_CONTEXT = FLOOR_REFERENCE_CONTEXT;
exports.ZONE_REFERENCE_CONTEXT = ZONE_REFERENCE_CONTEXT;
exports.ROOM_REFERENCE_CONTEXT = ROOM_REFERENCE_CONTEXT;
