// spinal-env-viewer-plugin-bimobjectservice/build/createGraphBimObject.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

let createGraph = (() => {
  var _ref = _asyncToGenerator(function* () {
    const forgeFile = yield window.spinal.spinalSystem.getModel();

    if (!forgeFile.hasOwnProperty("graph")) {
      forgeFile.add_attr({
        graph: new _spinalModelGraph.SpinalGraph()
      });
    }

    return forgeFile.graph;
  });

  return function createGraph() {
    return _ref.apply(this, arguments);
  };
})();

let createContext = (() => {
  var _ref2 = _asyncToGenerator(function* () {
    const graph = yield this.getGraph();
    let context = yield graph.getContext(BIM_OBJECT_CONTEXT_TYPE);

    if (context === undefined) {
      context = new _spinalModelGraph.SpinalContext(BIM_OBJECT_CONTEXT_TYPE);
      yield graph.addContext(context);
    }

    return context;
  });

  return function createContext() {
    return _ref2.apply(this, arguments);
  };
})();

var _spinalModelGraph = require("spinal-model-graph");

var _spinalModelsBimobject = require("spinal-models-bimobject");

var _spinalModelsBimobject2 = _interopRequireDefault(_spinalModelsBimobject);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const BIM_OBJECT_CONTEXT_TYPE = "BIMObjectContext";
const BIM_OBJECT_NODE_TYPE = "BIMObject";
const BIM_OBJECT_RELATION_NAME = "hasBIMObject";
const REFERENCE_OBJECT_RELATION_NAME = "hasReferenceObject";
const BIM_OBJECT_RELATION_TYPE = _spinalModelGraph.SPINAL_RELATION_PTR_LST_TYPE;

const bimObjectService = {
  graph: null,
  context: null,
  getGraph() {
    if (this.graph === null) {
      this.graph = createGraph();
    }

    return this.graph;
  },
  getContext() {
    if (this.context === null) {
      this.context = createContext.call(this);
    }

    return this.context;
  },
  createBIMObject(dbid, name) {
    var _this = this;

    return _asyncToGenerator(function* () {
      let BIMObjNode = yield _this.getBIMObject(dbid);

      if (BIMObjNode === undefined) {
        const BIMObject = new _spinalModelsBimobject2.default(dbid, name);

        BIMObjNode = new _spinalModelGraph.SpinalNode(name, BIM_OBJECT_NODE_TYPE, BIMObject);
        BIMObjNode.info.add_attr({
          dbid: dbid
        });

        const BIMObjectContext = yield _this.getContext();

        yield BIMObjectContext.addChildInContext(BIMObjNode, BIM_OBJECT_RELATION_NAME, BIM_OBJECT_RELATION_TYPE, BIMObjectContext);
      }

      return BIMObjNode;
    })();
  },
  getBIMObject(dbid) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      const BIMObjectContext = yield _this2.getContext(BIM_OBJECT_CONTEXT_TYPE);
      const BIMObjectArray = yield BIMObjectContext.getChildren([BIM_OBJECT_RELATION_NAME]);

      for (let BIMObject of BIMObjectArray) {
        if (BIMObject.info.dbid.get() === dbid) {
          return BIMObject;
        }
      }
    })();
  },
  addBIMObject(context, parent, dbId, name) {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      let node;

      if (dbId instanceof _spinalModelGraph.SpinalNode) {
        node = dbId;
      } else {
        node = yield _this3.getBIMObject(dbId);

        if (node === undefined) {
          node = yield _this3.createBIMObject(dbId, name);
        }
      }

      yield parent.addChildInContext(node, BIM_OBJECT_RELATION_NAME, BIM_OBJECT_RELATION_TYPE, context);

      return node;
    })();
  },
  removeBIMObject(parent, child) {
    return parent.removeChild(child, BIM_OBJECT_RELATION_NAME, BIM_OBJECT_RELATION_TYPE);
  },
  deleteBIMObject(dbId) {
    var _this4 = this;

    return _asyncToGenerator(function* () {
      const context = yield _this4.getContext();
      const children = yield context.getChildrenInContext();
      const child = children.find(function (node) {
        return node.info.dbId === dbId;
      });

      if (child === undefined) {
        throw Error("The dbId has no BIM object");
      }

      return child.removeFromGraph();
    })();
  },
  addReferenceObject(parent, dbId, name) {
    var _this5 = this;

    return _asyncToGenerator(function* () {
      let node;

      if (dbId instanceof _spinalModelGraph.SpinalNode) {
        node = dbId;
      } else {
        node = yield _this5.getBIMObject(dbId);

        if (node === undefined) {
          node = yield _this5.createBIMObject(dbId, name);
        }
      }

      yield parent.addChild(node, REFERENCE_OBJECT_RELATION_NAME, BIM_OBJECT_RELATION_TYPE);

      return node;
    })();
  },
  removeReferenceObject(parent, child) {
    return parent.removeChild(child, REFERENCE_OBJECT_RELATION_NAME, BIM_OBJECT_RELATION_TYPE);
  }
};

bimObjectService.constants = {
  BIM_OBJECT_CONTEXT_TYPE,
  BIM_OBJECT_NODE_TYPE,
  BIM_OBJECT_RELATION_NAME,
  REFERENCE_OBJECT_RELATION_NAME,
  BIM_OBJECT_RELATION_TYPE
};

exports.default = bimObjectService;
