// spinal-env-viewer-plugin-excel-manager-service/dist/classes/CreateExcel.js
"use strict";
/*
 * Copyright 2020 SpinalCom - www.spinalcom.com
 *
 * This file is part of SpinalCore.
 *
 * Please read all of the following terms and conditions
 * of the Free Software license Agreement ("Agreement")
 * carefully.
 *
 * This Agreement is a legally binding contract between
 * the Licensee (as defined below) and SpinalCom that
 * sets forth the terms and conditions that govern your
 * use of the Program. By installing and/or using the
 * Program, you agree to abide by all the terms and
 * conditions stated or referenced herein.
 *
 * If you do not agree to abide by these terms and
 * conditions, do not demonstrate your acceptance and do
 * not install or use the Program.
 * You should have received a copy of the license along
 * with this file. If not, see
 * <http://resources.spinalcom.com/licenses.pdf>.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Excel = require("exceljs");
class CreateExcel {
    constructor(sheets, author) {
        this.workbook = new Excel.Workbook();
        this.workbook.created = new Date(Date.now());
        this.sheets = sheets;
    }
    createSheet() {
        this.sheets.forEach((argSheet) => __awaiter(this, void 0, void 0, function* () {
            let sheet = this.workbook.addWorksheet(argSheet.name, { properties: { tabColor: { argb: 'FFC0000' } } });
            sheet.state = 'visible';
            yield this.addHeader(sheet, argSheet.header);
            this.addRows(sheet, argSheet.rows);
        }));
    }
    addHeader(sheet, headers) {
        if (sheet.columns && sheet.columns.length > 0) {
            sheet.columns = [...sheet.columns, ...headers];
        }
        else {
            sheet.columns = headers;
        }
    }
    addRows(sheet, argRows) {
        let rows = Array.isArray(argRows) ? argRows : [argRows];
        rows.forEach(row => {
            const r = sheet.addRow(row);
        });
    }
    getWorkbook() {
        // console.log(this.workbook);
        return this.workbook.xlsx.writeBuffer();
    }
    getWorkbookInstance() {
        return this.workbook;
    }
}
exports.default = CreateExcel;
